mullvad-obfuscated-dns-proxy/src/config/mod.rs

```diff
@@ -0,0 +1,88 @@
+use std::{
+    io::Cursor,
+    net::{Ipv6Addr, SocketAddrV4},
+};
+
+use byteorder::{LittleEndian, ReadBytesExt};
+
+mod plain;
+mod xor;
+pub use plain::Plain;
+pub use xor::Xor;
+
+#[derive(Debug)]
+pub enum Error {
+    UnknownType(u16),
+    InvalidXor(xor::Error),
+    InvalidPlain(plain::Error),
+}
+
+enum ProxyType {
+    Plain,
+    XorV1,
+    XorV2,
+}
+
+impl TryFrom<Ipv6Addr> for ProxyType {
+    type Error = Error;
+
+    fn try_from(value: Ipv6Addr) -> Result<Self, Self::Error> {
+        let mut data = Cursor::new(value.octets());
+        // skip the first 2 bytes since it's just padding to make the IP look more like a legit
+        // IPv6 address.
+        let _ = data
+            .read_u16::<LittleEndian>()
+            .expect("IPv6 must have at least 16 bytes");
+
+        match data
+            .read_u16::<LittleEndian>()
+            .expect("IPv6 must have at least 16 bytes")
+        {
+            0x01 => Ok(Self::Plain),
+            0x02 => Ok(Self::XorV1),
+            0x03 => Ok(Self::XorV2),
+            unknown => Err(Error::UnknownType(unknown)),
+        }
+    }
+}
+
+impl TryFrom<Vec<Ipv6Addr>> for AvailableProxies {
+    type Error = Error;
+
+    fn try_from(ips: Vec<Ipv6Addr>) -> Result<Self, Self::Error> {
+        let mut proxies = AvailableProxies {
+            plain: vec![],
+            xor: vec![],
+        };
+
+        for ip in ips {
+            match ProxyType::try_from(ip)? {
+                ProxyType::Plain => {
+                    proxies
+                        .plain
+                        .push(Plain::try_from(ip).map_err(Error::InvalidPlain)?);
+                }
+                ProxyType::XorV2 => {
+                    proxies
+                        .xor
+                        .push(Xor::try_from(ip).map_err(Error::InvalidXor)?);
+                }
+                // this type is ignored.
+                ProxyType::XorV1 => continue,
+            }
+        }
+
+        Ok(proxies)
```

Comment:
Very much a style preference, but it seems you could remove the `ProxyType` enum entirely and use this
```suggestion
        let plain = ips
            .iter()
            .filter_map(|ip| Plain::try_from(*ip).ok())
            .collect();
        let xor = ips
            .iter()
            .filter_map(|ip| Xor::try_from(*ip).ok())
            .collect();
        Ok(AvailableProxies { plain, xor })
```

Reply:
I took a look at this again, and I wonder if maybe the simplest implementation would be to also remove `Plain::try_from` and `Xor::try_from` and put all the logic in the `AvailableProxies::try_from` implementation like this.

```rust
#[derive(Debug)]
pub enum Error {
    UnknownType(u16),
    InvalidXor(xor::Error),
    InvalidPlain(plain::Error),
    UnexpectedType(u16),
    EmptyXorKey,
}

impl TryFrom<Vec<Ipv6Addr>> for AvailableProxies {
    type Error = Error;

    fn try_from(ips: Vec<Ipv6Addr>) -> Result<Self, Self::Error> {
        let mut proxies = AvailableProxies {
            plain: vec![],
            xor: vec![],
        };

        for ip in ips {
            let mut data = Cursor::new(ip.octets());
            // skip the first 2 bytes since it's just padding to make the IP look more like a
            // legit IPv6 address.

            data.set_position(2);
            let proxy_type = data
                .read_u16::<LittleEndian>()
                .expect("IPv6 must have at least 16 bytes");
            match proxy_type {
                0x01 => {
                    let mut ipv4_bytes = [0u8; 4];
                    data.read_exact(&mut ipv4_bytes).unwrap();
                    let v4_addr = Ipv4Addr::from(ipv4_bytes);

                    let port = data.read_u16::<LittleEndian>().unwrap();
                    proxies.plain.push(Plain {
                        addr: SocketAddrV4::new(v4_addr, port),
                    });
                }
                0x02 => {
                    let mut ipv4_bytes = [0u8; 4];
                    data.read_exact(&mut ipv4_bytes).unwrap();
                    let v4_addr = Ipv4Addr::from(ipv4_bytes);

                    let port = data.read_u16::<LittleEndian>().unwrap();

                    let mut key_bytes = [0u8; 6];
                    data.read_exact(&mut key_bytes).unwrap();
                    let xor_key = key_bytes
                        .into_iter()
                        .filter(|byte| *byte != 0x00)
                        .collect::<Vec<_>>();
                    if xor_key.is_empty() {
                        return Err(Error::EmptyXorKey);
                    }
                    proxies.xor.push(Xor {
                        addr: SocketAddrV4::new(v4_addr, port),
                        xor_key,
                        key_index: 0,
                    });
                }
                // this type is ignored.
                0x03 => continue,
                unknown => return Err(Error::UnknownType(unknown)),
            }
        }

        Ok(proxies)
    }
}
```

It would remove the duplicate checks of the proxy type byte and some other boilerplate.